### PR TITLE
Added `compileWithPayload` combinator

### LINKED ISSUE #370: A function that transforms (part of) inputs into extra witness

### PR DESCRIPTION
**Pull request by TurtlePU**
Resolves #370 by adding a new flavor of `compile` function which takes inputs from the payload instead of usual input variables

### CODE REVIEW

File: symbolic-base/src/ZkFold/Symbolic/Compiler.hs [hunk 1]
```diff
@@ -44,85 +47,86 @@ import           ZkFold.Symbolic.MonadCircuit               (MonadCircuit (..))
...
+  forall a p c f s l y .
+  ( c ~ ArithmeticCircuit a l p, Arithmetic a, Binary a
+  , Binary (Rep p), Ord (Rep p), NFData (Rep p)
+  , Traversable l
+  , SymbolicData f, Context f ~ c, Support f ~ s
+  , SymbolicInput s, Context s ~ c, Layout s ~ l
+  , SymbolicData y, Context y ~ c, Support y ~ Proxy c
+  , Layout f ~ Layout y
+  ) => f -> y
+compileWithPayload = compileCircuit . solderWith payloadCircuit
```
Comment (vlasin) [inline symbolic-base/src/ZkFold/Symbolic/Compiler.hs:108]: Proposal: let's consider having a single `compile` function with an additional "options" argument. Would it help avoid the mistake of selecting a "wrong" compile function?

Comment (TurtlePU) [reply symbolic-base/src/ZkFold/Symbolic/Compiler.hs]: Well, we can leave only the `compileWith f` function which would look like `compileCircuit . f`, where `f` can be a `solderWith c` postcomposed with some desired operations

Comment (TurtlePU) [reply symbolic-base/src/ZkFold/Symbolic/Compiler.hs]: Left only `solderWith`, `finalRestore`, `compile` and `compileIO` in the public API

File: symbolic-base/src/ZkFold/Symbolic/Compiler.hs [hunk 2]
```diff
@@ -41,88 +37,50 @@ import           ZkFold.Symbolic.MonadCircuit               (MonadCircuit (..))
...
-    , Layout f ~ Layout y
-    , Symbolic c
-    ) => f -> y
-compile = restore . const . optimize . solder @a
+compile :: forall a y f c s p.
+  ( SymbolicData f, Context f ~ c, Support f ~ s, SymbolicInput s, Context s ~ c
+  , c ~ ArithmeticCircuit a p (Layout s), Arithmetic a, Binary a, Binary (Rep p)
+  , SymbolicData y, Context y ~ c, Support y ~ Proxy c, Layout y ~ Layout f)
+  => f -> y
+compile = finalRestore . solderWith idCircuit
```
Comment (vlasin) [inline symbolic-base/src/ZkFold/Symbolic/Compiler.hs:62]: I propose adding `compileWith` function that has parameters: a circuit for `solderWith` and a transformation after the `solderWith` function (like `forceOne` or `guessOutput`).

**Review by vlasin — APPROVED** [review]

**Pull request merged**